api/pin.go

```
@@ -0,0 +1,426 @@
+// uploading the file using the pin command. This function can pin both
+// encrypted and non-encrypted files.
+func (p *PinAPI) PinFiles(rootHash string, isRaw bool, credentials string) error {
+
+	addr, err := hex.DecodeString(rootHash)
+	if err != nil {
+		log.Error("Error decoding root hash" + err.Error())
+		return err
+	}
+
+	hasCHunk, err := p.db.Has(context.TODO(), chunk.Address(p.removeDecryptionKeyFromChunkHash(addr)))
+	if !hasCHunk {
+		log.Error("Could not pin hash. File not uploaded", "Hash", rootHash)
+		return err
+	}
+
+	// Walk the root hash and pin all the chunks
+	walkerFunction := func(ref storage.Reference) error {
+		chunkAddr := p.removeDecryptionKeyFromChunkHash(ref)
+		err := p.db.Set(context.TODO(), chunk.ModeSetPin, chunkAddr)
+		if err != nil {
+			log.Error("Could not pin chunk. Address " + hex.EncodeToString(chunkAddr))
+			return err
+		} else {
+			log.Trace("Pinning chunk", "Address", hex.EncodeToString(chunkAddr))
+		}
+		return nil
+	}
+	p.walkChunksFromRootHash(rootHash, isRaw, credentials, walkerFunction)
+
+	// Check if the root hash is already pinned
+	isFilePinned := p.db.IsFilePinned(chunk.Address(addr))
+	if !isFilePinned {
+		if isRaw {
+			err = p.db.Set(context.TODO(), chunk.ModeSetRawFile, addr)
+		} else {
+			err = p.db.Set(context.TODO(), chunk.ModeSetFile, addr)
+		}
+		if err != nil {
+			// TODO: if this happens, we should go back and revert the entire file's chunks
+			log.Error("Could not unpin root chunk. Address " + fmt.Sprintf("%0x", addr))
+			return err
+		}
+	}
+	return nil
+}
+
+// UnPinFiles is used to unpin an already pinned file. It takes the root
+// hash of the file and walks down the merkle tree unpinning all the chunks
+// that are encountered on the way. The pre-requisite is that the file should
+// have been already pinned using the PinFiles function. This function can
+// be called only from an external command.
+func (p *PinAPI) UnpinFiles(rootHash string, credentials string) {
```

why doesn't this function return an error? this way the calling context can know an error occurred. please return the errors below and handle them in the calling context

---

No one calls them.. its a command called from websocket. 
Anyway it returns error now.